android/src/main/java/com/reactlibrary/RNCarnivalModule.java

```diff
@@ -40,524 +38,518 @@
-    AttributeMap carnivalAttributeMap = new AttributeMap();
-    carnivalAttributeMap.setMergeRules(attributeMap.getInt("mergeRule"));
-
-    Iterator<String> keys = attributes.keys();
-
-    while (keys.hasNext()) {
-      String key = keys.next();
-      JSONObject attribute = attributes.getJSONObject(key);
-      String attributeType = attribute.getString("type");
-      if (attributeType.equals("string")) {
-        carnivalAttributeMap.putString(key, attribute.getString("value"));
-
-      } else if (attributeType.equals("stringArray")) {
-        ArrayList<String> array = new ArrayList<String>();
-        JSONArray values = attribute.getJSONArray("value");
-        for (int i = 0; i < values.length(); i++) {
-          array.add((String)values.get(i));
-        }
-
-        carnivalAttributeMap.putStringArray(key, array);
+public class RNCarnivalModule extends ReactContextBaseJavaModule implements Carnival.OnInAppNotificationDisplayListener {
 
-      } else if (attributeType.equals("integer")) {
-        carnivalAttributeMap.putInt(key, attribute.getInt("value"));
+    protected final static String ERROR_CODE_DEVICE = "carnival.device";
+    protected final static String ERROR_CODE_MESSAGES = "carnival.messages";
+    protected final static String ERROR_CODE_RECOMMENDATIONS = "carnival.recommendations";
+    protected final static String ERROR_CODE_TRACKING = "carnival.tracking";
 
-      } else if (attributeType.equals("integerArray")) {
-        ArrayList<Integer> array = new ArrayList<Integer>();
-        JSONArray values = attribute.getJSONArray("value");
-        for (int i = 0; i < values.length(); i++) {
-          Object j = values.getInt(i);
-          array.add((Integer)j);
-        }
-
-        carnivalAttributeMap.putIntArray(key, array);
+    private boolean displayInAppNotifications;
 
-      } else if (attributeType.equals("boolean")) {
-        carnivalAttributeMap.putBoolean(key, attribute.getBoolean("value"));
+    private ReactApplicationContext reactApplicationContext;
 
-      } else if (attributeType.equals("float")) {
-        carnivalAttributeMap.putFloat(key, (float)attribute.getDouble("value"));
+    public RNCarnivalModule(ReactApplicationContext reactContext, boolean displayInAppNotifications) {
+        super(reactContext);
+        reactApplicationContext = reactContext;
+        this.displayInAppNotifications = displayInAppNotifications;
 
-      } else if (attributeType.equals("floatArray")) {
-        ArrayList<Float> array = new ArrayList<Float>();
-        JSONArray values = attribute.getJSONArray("value");
-        for (int i = 0; i < values.length(); i++) {
-          Float value = Float.parseFloat(values.get(i).toString());
-          array.add(value);
-        }
+        Carnival.setOnInAppNotificationDisplayListener(this);
+        setWrapperInfo();
+    }
 
-        carnivalAttributeMap.putFloatArray(key, array);
+    @Override
+    public boolean shouldPresentInAppNotification(Message message) {
```

Comment:
I still not understanding why our `toJSON` method is protected. Why not just make it public on our SDK? (optional)

Reply:
Yeah, we could take the Message class down the same route as the ContentItem class and just make it public. I might look into that for a later SDK release as it would remove some need for reflection.